---
title: "Active subnetwork detection Quick Start"
author: "Yubing Yao and Rebekah Kristal"
date: '2023-05-17'
output: 
  rmdformats::robobook:
    code_folding: show
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Install and Load the `wActNet` Package

Begin by installing and loading the `devtools` package. Use the install_github function to install `wActNet` from `GitHub` as shown below. Our package depends on two additional R packages, `igraph` and `BioNet`, and the second one can be downloaded from `BioConductor` <https://www.bioconductor.org/install/>.


```{r,results='hide' }
#please comment the below code out if you already install the package-wActNet from github
#library(devtools)
#if (!requireNamespace("BiocManager", quietly = TRUE))
#  install.packages("BiocManager")
#BiocManager::install("pathview")
#install_github("yuy113/wActNet")

library(wActNet)
#R package BioNet from BioConductor 
library(BioNet)
library(igraph)
#real data set from 

```
## Simulated dataset
### Simulate data and generate P values of tests on pairwise correlations


To illustrate the process, we simulate data ($n$ observations, $p$ features) from a multivariate normal distribution, which is then used to generate data on ${p}\choose{2}$ pairwise correlations data through the simulation of random data, which follows a multivariate normal distribution. Subsequently, we use the function `pval.perm.corr()` to conduct multiple conduct nonparametric permutation tests for all the pairwise correlations within the simulated data. Using the function and obtaining corresponding P values, we obtain the corresponding P-values from these permutation tests.



```{r}
dat1<-matrix(rnorm(4000),ncol=40,nrow=100)
colnames(dat1)<-paste("Var", as.character(1:40),sep="")
pval.edge<-pval.perm.corr(dat1,nsim=100,MatchId=NULL,do.parallel=FALSE)
#pval.edge
```

### Calculate node scores and edge scores 


To illustrate our approach, we begin by simulating a network of 40 nodes. We generate P values for all nodes and potential connecting edges using the `uniform.beta.node.edge.score()` function. The output of this function is a list consisting of node scores (`NodeScore`), edge scores (`EdgeScore`), and the network data in the form of an igraph object (`Network`).



```{r}
# simulate the p values for all the possible edges in the network
ind.pos.pval.edge<-rbinom(40*39/2,1,0.5)
pval.edge<-(1-ind.pos.pval.edge)*runif(40*39/2)+ind.pos.pval.edge*rbeta(40*39/2,0.1,1)
names(pval.edge)<-unlist(sapply(1:39,function(i){sapply((i+1):40, function(j){paste(paste("Var",
as.character(i),sep=""),paste("Var",as.character(j),sep=""),sep="_")})}))

# simulate p values for all the nodes in the network
ind.pos.pval.node<-rbinom(40,1,0.2)
pval.node<-(1-ind.pos.pval.node)*runif(40)+ind.pos.pval.node*rbeta(40,0.1,1)
names(pval.node)<-paste("Var", as.character(1:40),sep="")

# generate the node score-NodeScore, edge scores-EdgeScore and igraph object-Network
network.test<-uniform.beta.node.edge.score(pval.node,pval.edge,0.05,0.05,dat1)

#igraph object-Network
network.test$Network
#node scores
network.test$NodeScore
#edge scores
#network.test$EdgeScore
```


### Detect optimized subnetworks 


To detect active subnetworks within our network, we illustrate the approach by Dittrich et al. in 2008 and compare to the `wActNet` algorithm. The algorithm proposed by Dittrich et al. in 2008 uses node scores only. The `wActNet` algorithm combines node scores and edge scores using the `MultiModuleFind` function. Within the `MultiModuleFind` function, we have the flexibility to set the maximum number of active subnetworks (`ncluster`) and choose between two methods: the one presented by Dittrich et al. in 2008 (method=`NodeOnly`) or the wActNet approach (method=`NodeEdge`). 

Furthermore, the wActNet algorithm allows us to fine-tune the optimization of active subnetworks by adjusting the parameter `weightratio.edge.node`. This parameter represents the weighted ratio between edge scores and node scores in the objective function.

In summary, by combining the methodology introduced by Dittrich et al. (2008) with the `wActNet` extension that includes both  node and edge scores, we can effectively detect the optimized active subnetworks on network data.


Please note when the function completes execution and fails to identify any further optimized or active subnetworks, a warning message "No positive nodes" will be generated.


```{r}
network.test1<-network.test$Network
node.scores<-network.test$NodeScore
edge.scores<-network.test$EdgeScore

#Note: both BioNet and wActNet R packages stop running and can't detect any optimized/active subnetwork further and complete, a warning output message-"No positive nodes" will be generated.
#identify all possible optimized subnetworks in simulated network
# using Dittrich's method using node scores only
multi.mod.n<-MultiModuleFind(network.test1,node.scores,edge.scores,
weightratio.edge.node=1,ncluster=3,method="NodeOnly")
#all possible optimized subnetworks in simulated network using Dittrich's method using node scores only
multi.mod.n
#visualize the final illustrated detected optimized subnetwork by BioNet algorithm
#for edges, edge score >=0, solid line, edge score < 0, dash line
#for nodes, node score >=0, red color, edge score < 0, grey color 
#the size/length of node or edges proportional to node or edge score respectively
#need R package-visNetwork, if not, please install from CRAN
require(visNetwork, quietly = TRUE)
dat_node<-multi.mod.n[[1]]
 E(dat_node)$dashes<-ifelse(E(dat_node)$score>0,FALSE,FALSE)
  E(dat_node)$width<-(E(dat_node)$score+1.5)*0.3
  V(dat_node)$color<-ifelse(V(dat_node)$score>0,"#E7298A","grey")
  V(dat_node)$size<-(V(dat_node)$score+11)*0.5
  
  E(dat_node)$color<-ifelse(E(dat_node)$score>0,"#66A61E","grey")
  
  
 #########################################################################################
  # setwd("/Users/yubingyao/Google Drive/Network analysis/R code/")
  
 # save(GU2,file="network2_2clusters_plot.RData")
  
  
  
  #load("network_2clusters_plot.RData")
  
  data_node<- toVisNetworkData(dat_node)
  
  visNetwork(nodes = data_node$nodes, edges = data_node$edges, height = "800px")%>% visNodes(font=list(size=20))
# identify all possible signaling modules in simulated network
# using our proposed method based on node scores and edge scores
multi.mod.e<-MultiModuleFind(network.test1,node.scores,edge.scores,
weightratio.edge.node=1,ncluster=3,method="NodeEdge")
#all possible optimized subnetworks in simulated network using our proposed method based on node scores and edge scores
multi.mod.e

#visualize the final illustrated detected optimized subnetwork by our proposed algorithm
#for edges, edge score >=0, solid line, edge score < 0, dash line
#for nodes, node score >=0, red color, edge score < 0, grey color 
#the size/length of node or edges proportional to node or edge score respectively
#need R package-visNetwork, if not, please install from CRAN
#require(visNetwork, quietly = TRUE)
dat_nodeedge<-multi.mod.e[[1]]
 E(dat_nodeedge)$dashes<-ifelse(E(dat_nodeedge)$score>0,FALSE,FALSE)
  E(dat_nodeedge)$width<-(E(dat_nodeedge)$score+1.5)*0.3
  V(dat_nodeedge)$color<-ifelse(V(dat_nodeedge)$score>0,"#E7298A","grey")
  V(dat_nodeedge)$size<-(V(dat_nodeedge)$score+11)*0.5
  
  E(dat_nodeedge)$color<-ifelse(E(dat_nodeedge)$score>0,"#66A61E","grey")
  
  
 #########################################################################################
  # setwd("/Users/yubingyao/Google Drive/Network analysis/R code/")
  
 # save(GU2,file="network2_2clusters_plot.RData")
  
  
  
  #load("network_2clusters_plot.RData")
  
  data_nodeedge <- toVisNetworkData(dat_nodeedge)
  
  visNetwork(nodes = data_nodeedge$nodes, edges = data_nodeedge$edges, height = "800px")%>% visNodes(font=list(size=20))
```
## Real dataset of ovarian cancer
This uses a set of public ovarian cancer data in the curatedOvarianData R package. See "??curatedOvarianData" in the RStudio console for more help with this dataset: "This package represents a manually curated data collection for gene expression meta-analysis of patients with ovarian cancer. This resource provides uniformly prepared microarray data with curated and documented clinical metadata. It allows a computational user to efficiently curatedOvarianData identify studies and patient subgroups of interest for analysis and to run such analyses immediately without the challenges posed by harmonizing heterogeneous microarray technologies, study designs, expression data processing methods, and clinical data formats."

Because this is full-transcriptome data, we will load the data and select only a subset of it.
### Downloading and preprocessing ovarian cancer data

```{R installAndLoadPackages, cache=T}

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
#BiocManager::install(version = "3.19")

library(BiocManager)
# Uncomment this line to install the affy and curatedOvarianData package 
# Takes time, so re-comment after
# BiocManager::install("affy")
# BiocManager::install("curatedOvarianData")

#library(devtools)
#install_github("yuy113/wActNet", force = TRUE)
library(igraph)

library(affy)
#library(BioNet)
library(dplyr)
library(stringr)
#library(wActNet)
library(curatedOvarianData)

```


###Load and select a subset of the ovarian cancer data.

```{R loadOcData}
standardize = function(x){return((x-mean(x))/sd(x))}

data(GSE32062.GPL6480_eset)
lateStage = exprs(GSE32062.GPL6480_eset)
```

For a straightforward example on a small set of genes, use the chunk below:

```{R selectHPOGenes}
# Extract a subset of genes related to ovarian carcinoma 
# based on the Human Phenotype Ontology
# See https://hpo.jax.org/app/browse/term/HP:0025318

lateStageSmall = lateStage[c(1680,1681,3027,4564,8930,12243,12245,13694,13695,13701,13979,16082,16875,17980),] %>%
  t() %>%
  data.frame() %>%
  apply(.,2,standardize)
```
Select a set of 50 genes/predictors
```{R increasePredictorSize, eval=T}
# select a random subset of genes to test different sizes
setSize = 50
set.seed(1989)
# remove genes with weird symbols in the names as this messes up wActNet
lateStage = lateStage[-grep("///",row.names(lateStage)),]
geneIdx = sample(1:ncol(lateStage), size=setSize,replace=F)
lateStageSmall = lateStage[geneIdx,] %>%
  t() %>%
  data.frame() %>%
  apply(.,2,standardize)
names(lateStageSmall) = colnames(lateStageSmall)
head(lateStageSmall)
```


### Assigning edge p-values with permutation on a correlation network

To generate edge p-values, we use a permutation test on a correlation network. You can vary `nsim` to make this easier to test, although in practice you'll want to use a high number. 

```{R}
edge_p_vals = pval.perm.corr(dat = lateStageSmall,MatchId = NULL,
               nsim = 1000, do.parallel = F, no_cores = 1)


```


### Assigning node p-values associated with covariate

To assign node p-values, we will regress the expression of each gene against summary grade.

```{R}
lateStagePheno = phenoData(GSE32062.GPL6480_eset)
gradeInfo = data.frame("sample"=row.names(pData(lateStagePheno)),
                       "summarygrade"=pData(lateStagePheno)$summarygrade) 
regressionDataset = lateStageSmall %>% 
  data.frame() %>%
  mutate("sample"=row.names(lateStageSmall)) %>%
  inner_join(gradeInfo,by="sample")

```
To get node weights, we regress gene expression against summary grade and take the resulting p-values. Note that we take the raw p-values and not the FDR, as the conversion to node scores in wActNet provides FDR control.

```{R nodeRegressions}
library(limma)
design_matrix = model.matrix( ~ factor(summarygrade),data=regressionDataset)
limma_res = regressionDataset %>% select(-c(sample,summarygrade)) %>% t() %>%
  lmFit(object = ., design = design_matrix) %>%
  eBayes()
node_table = limma_res %>%
  topTable(number = Inf) %>%
  data.frame()  
node_p_vals = node_table$P.Value
names(node_p_vals) = rownames(node_table)
```


### Running wActNet

#### Converting p-values to node scores

The `uniform.beta.node.edge.score` function is provided in `wActNet` for converting the list of p-values to scores, in which a higher score corresponds to a lower p-value and vice-versa. FDR control is implemented such that negative scores correspond to p-values above the FDR threshold and positive scores correspond to p-values that are FDR-significant (i.e., at or below the FDR threshold).  

```{R}
networkPrep = uniform.beta.node.edge.score(pval.node = node_p_vals,
                                           pval.edge = edge_p_vals,
                                           FDR.node=0.1,
                                           FDR.edge=0.05,
                                           dat =data.frame(lateStageSmall) )


lateStageSmall1<-data.frame(lateStageSmall)

networkPrep$NodeScore
node_p_vals[order(node_p_vals)]
any(is.na(names(edge_p_vals)))
names(node_p_vals)
names(lateStageSmall)<-as.vector(na.omit(names(lateStageSmall)))
```

#### Running module detection

The `MultiModuleFind` function is used to find active subnetworks. It is an iterative algorithm that finds the largest scoring subnetwork, then removes it from the network and repeats the process. The number of modules `ncluster` must be prespecified. The algorithm will stop before reaching `ncluster` if there are no positive nodes left in the network.

```{R}
modules = MultiModuleFind(network = networkPrep$Network,
                          node.scores = networkPrep$NodeScore,
                          edge.scores = networkPrep$EdgeScore,
                          ncluster = 2,
                          method = "NodeEdge")



```

Examine results:

```{R}
modules
```

## Contact Us / Contribute
This package is new, and any and all suggestions are welcomed. You can use GitHub to raise issues, contribute, or communicate with us about the package:

https://github.com/yuy113/wActNet/

## Reference

Dittrich, M., Klau, G., Rosenwald, A., Dandekar, T., Müller, T.: Identifying functional modules in protein-protein interaction
networks: an integrated exact approach. Bioinformatics 24(13), 223–231 (2008)

Beisser, D., Klau, G., Dandekar, T., Müller, T., Dittrich, M.: Bionet: an r-package for the functional analysis of biological
networks. Bioinformatics 26(8), 1129–1130 (2010)
