---
title: "Active subnetwork detection Quick Start"
author: "Yubing Yao, Rebekah Kristal and Katherine H. Shutta"
date: '2024-10-09'
output: 
  rmdformats::robobook:
    code_folding: show
    toc_depth: 5
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Install and Load the `wActNet` Package

Begin by installing and loading the `devtools` package. Use the install_github function to install `wActNet` from `GitHub` as shown below. Our package depends on two additional R packages, `igraph` and `BioNet`, and the second one can be downloaded from `BioConductor` <https://www.bioconductor.org/install/>.


```{r,results='hide' }
#please comment the below code out if you already install the package-wActNet from github
#library(devtools)
#if (!requireNamespace("BiocManager", quietly = TRUE))
#  install.packages("BiocManager")
#BiocManager::install("pathview")
#install_github("yuy113/wActNet")
#BiocManager::install(c("BioNet"))
#foreign, MASS, nlme, survival
library(wActNet)
#R package BioNet from BioConductor 
library(BioNet)
library(igraph)


```
## Simulated dataset

### Simulate data and generate P values of tests on pairwise correlations


To illustrate the process, we simulate data ($n$ observations, $p$ features) from a multivariate normal distribution, which is then used to generate data on ${p}\choose{2}$ pairwise correlations data through the simulation of random data, which follows a multivariate normal distribution. Subsequently, we use the function `pval.perm.corr()` to conduct multiple  nonparametric permutation tests for all the pairwise correlations within the simulated data. Using the function and obtaining corresponding P values, we obtain the corresponding P-values from these permutation tests.



```{r}
dat1<-matrix(rnorm(4000),ncol=40,nrow=100)
colnames(dat1)<-paste("Var", as.character(1:40),sep="")
pval.edge<-pval.perm.corr(dat1,nsim=100,MatchId=NULL,do.parallel=FALSE)
#pval.edge
```

### Calculate node scores and edge scores 


To illustrate our approach, we begin by simulating a network of 40 nodes. We generate P values for all nodes and potential connecting edges using the `uniform.beta.node.edge.score()` function. The output of this function is a list consisting of node scores (`NodeScore`), edge scores (`EdgeScore`), and the network data in the form of an igraph object (`Network`).



```{r}
# simulate the p values for all the possible edges in the network
ind.pos.pval.edge<-rbinom(40*39/2,1,0.5)
pval.edge<-(1-ind.pos.pval.edge)*runif(40*39/2)+ind.pos.pval.edge*rbeta(40*39/2,0.1,1)
names(pval.edge)<-unlist(sapply(1:39,function(i){sapply((i+1):40, function(j){paste(paste("Var",
as.character(i),sep=""),paste("Var",as.character(j),sep=""),sep="_")})}))

# simulate p values for all the nodes in the network
ind.pos.pval.node<-rbinom(40,1,0.2)
pval.node<-(1-ind.pos.pval.node)*runif(40)+ind.pos.pval.node*rbeta(40,0.1,1)
names(pval.node)<-paste("Var", as.character(1:40),sep="")

# generate the node score-NodeScore, edge scores-EdgeScore and igraph object-Network
network.test<-uniform.beta.node.edge.score(pval.node,pval.edge,0.05,0.05,dat1)

#igraph object-Network
network.test$Network
#node scores
network.test$NodeScore
#edge scores
#network.test$EdgeScore
```


### Detect optimized subnetworks 


To detect active subnetworks within our network, we illustrate the approach by Dittrich et al. in 2008 and compare to the `wActNet` algorithm. The algorithm proposed by Dittrich et al. in 2008 uses node scores only. The `wActNet` algorithm combines node scores and edge scores using the `MultiModuleFind` function. Within the `MultiModuleFind` function, we have the flexibility to set the maximum number of active subnetworks (`ncluster`) and choose between two methods: the one presented by Dittrich et al. in 2008 (method=`NodeOnly`) or the wActNet approach (method=`NodeEdge`). 

Furthermore, the wActNet algorithm allows us to fine-tune the optimization of active subnetworks by adjusting the parameter `weightratio.edge.node`. This parameter represents the weighted ratio between edge scores and node scores in the objective function.

In summary, by combining the methodology introduced by Dittrich et al. (2008) with the `wActNet` extension that includes both  node and edge scores, we can effectively detect the optimized active subnetworks on network data.


Please note when the function completes execution and fails to identify any further optimized or active subnetworks, a warning message "No positive nodes" will be generated.


```{r}
network.test1<-network.test$Network
node.scores<-network.test$NodeScore
edge.scores<-network.test$EdgeScore

#Note: both BioNet and wActNet R packages stop running and can't detect any optimized/active subnetwork further and complete, a warning output message-"No positive nodes" will be generated.
#identify all possible optimized subnetworks in simulated network
# using Dittrich's method using node scores only
multi.mod.n<-MultiModuleFind(network.test1,node.scores,edge.scores,
weightratio.edge.node=1,ncluster=3,method="NodeOnly")
#all possible optimized subnetworks in simulated network using Dittrich's method using node scores only
multi.mod.n
#visualize the final illustrated detected optimized subnetwork by BioNet algorithm
#for edges, edge score >=0, solid line, edge score < 0, dash line
#for nodes, node score >=0, red color, edge score < 0, grey color 
#the size/length of node or edges proportional to node or edge score respectively
#need R package-visNetwork, if not, please install from CRAN
require(visNetwork, quietly = TRUE)
dat_node<-multi.mod.n[[1]]
 E(dat_node)$dashes<-ifelse(E(dat_node)$score>0,FALSE,FALSE)
  E(dat_node)$width<-(E(dat_node)$score+1.5)*0.3
  V(dat_node)$color<-ifelse(V(dat_node)$score>0,"#E7298A","grey")
  V(dat_node)$size<-(V(dat_node)$score+11)*0.5
  
  E(dat_node)$color<-ifelse(E(dat_node)$score>0,"#66A61E","grey")
  
  
 #########################################################################################
  # setwd("/Users/yubingyao/Google Drive/Network analysis/R code/")
  
 # save(GU2,file="network2_2clusters_plot.RData")
  
  
  
  #load("network_2clusters_plot.RData")
  
  data_node<- toVisNetworkData(dat_node)
  
  visNetwork(nodes = data_node$nodes, edges = data_node$edges, height = "800px")%>% visNodes(font=list(size=20))
# identify all possible signaling modules in simulated network
# using our proposed method based on node scores and edge scores
multi.mod.e<-MultiModuleFind(network.test1,node.scores,edge.scores,
weightratio.edge.node=1,ncluster=3,method="NodeEdge")
#all possible optimized subnetworks in simulated network using our proposed method based on node scores and edge scores
multi.mod.e

#visualize the final illustrated detected optimized subnetwork by our proposed algorithm
#for edges, edge score >=0, solid line, edge score < 0, dash line
#for nodes, node score >=0, red color, edge score < 0, grey color 
#the size/length of node or edges proportional to node or edge score respectively
#need R package-visNetwork, if not, please install from CRAN
#require(visNetwork, quietly = TRUE)
dat_nodeedge<-multi.mod.e[[1]]
 E(dat_nodeedge)$dashes<-ifelse(E(dat_nodeedge)$score>0,FALSE,FALSE)
  E(dat_nodeedge)$width<-(E(dat_nodeedge)$score+1.5)*0.3
  V(dat_nodeedge)$color<-ifelse(V(dat_nodeedge)$score>0,"#E7298A","grey")
  V(dat_nodeedge)$size<-(V(dat_nodeedge)$score+11)*0.5
  
  E(dat_nodeedge)$color<-ifelse(E(dat_nodeedge)$score>0,"#66A61E","grey")
  
  
 #########################################################################################
  # setwd("/Users/yubingyao/Google Drive/Network analysis/R code/")
  
 # save(GU2,file="network2_2clusters_plot.RData")
  
  
  
  #load("network_2clusters_plot.RData")
  
  data_nodeedge <- toVisNetworkData(dat_nodeedge)
  
  visNetwork(nodes = data_nodeedge$nodes, edges = data_nodeedge$edges, height = "800px")%>% visNodes(font=list(size=20))
```
## Application on the dataset of ovarian cancer

This uses a set of public ovarian cancer data in the curatedOvarianData R package. See "??curatedOvarianData" in the RStudio console for more help with this dataset: "This package represents a manually curated data collection for gene expression meta-analysis of patients with ovarian cancer. This resource provides uniformly prepared microarray data with curated and documented clinical metadata. It allows a computational user to efficiently curatedOvarianData identify studies and patient subgroups of interest for analysis and to run such analyses immediately without the challenges posed by harmonizing heterogeneous microarray technologies, study designs, expression data processing methods, and clinical data formats."

As with the simulation above, begin by installing and loading the devtools package. Use the install_github function to install wActNet from GitHub as shown below. Our package depends on two additional R packages: igraph and BioNet. Additionally, install the curatedOvarianData package containing the data to be used in this application example.

Because this is full-transcriptome data, we will load the data and select only a subset of it.

### Install and load packages for downloading and preprocessing of data and for running wActNet

```{R installAndLoadPackages, cache=T}
start.time <- Sys.time()
library(devtools)


# Uncomment these lines to install the affy and curatedOvarianData package 
# Takes time, so re-comment after

# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install(version = "3.19")
# library(BiocManager)

# BiocManager::install("affy")
# BiocManager::install("curatedOvarianData")

library(Biobase)

#install the newest version of wActNet package
#install_github("yuy113/wActNet", force = TRUE)

library(igraph)


library(affy)
#library(BioNet)
library(dplyr)
library(stringr)
#library(wActNet)
library(curatedOvarianData)


```


### Load and select a subset of the ovarian cancer data.

```{R loadOcData}
standardize = function(x){return((x-mean(x))/sd(x))}

data(GSE32062.GPL6480_eset)
lateStage = exprs(GSE32062.GPL6480_eset)
```

For a straightforward example on a small set of genes, use the chunk below:

```{R selectHPOGenes}
# Extract a subset of genes related to ovarian carcinoma 
# based on the Human Phenotype Ontology
# See https://hpo.jax.org/app/browse/term/HP:0025318
library(tidyr)

library(tidyverse)
lateStageSmall = lateStage[c(1680,1681,3027,4564,8930,12243,12245,13694,13695,13701,13979,16082,16875,17980),] %>%
  t() %>%
  data.frame() %>%
  apply(.,2,standardize)
```
Use the code chunk below to instead randomly select a larger group of genes. Note that wActNet does not handle gene names with certain symbols, for example "///" in the ovarian cancer dataset. Additionally, underscores will not work because this symbol is used later in order to connect two node names. We provide an example workaround below:

```{R increasePredictorSize, eval=T}
# select a random subset of genes to test different sizes
setSize = 500
set.seed(1989)
# remove genes with weird symbols in the names as this messes up wActNet
row.names(lateStage) <- gsub("///", "-", row.names(lateStage))

geneIdx = sample(1:nrow(lateStage), size=setSize,replace=F)
lateStageSmall = lateStage[geneIdx,] %>%
  t() %>%
  data.frame() %>%
  apply(.,2,standardize)
names(lateStageSmall) = colnames(lateStageSmall)
#head(lateStageSmall)
```


### Assigning edge p-values 

To generate edge p-values, we use a permutation test on a correlation network. You can vary `nsim` to make this easier to test, although in practice you'll want to use a high number. 

```{R}
edge_p_vals = wActNet::pval.perm.corr(dat = lateStageSmall,MatchId = NULL,
               nsim = 100, do.parallel = F, no_cores = 1)

#names(edge_p_vals)

```


### Assigning node p-values 

To assign node p-values associated with each covariate, we will regress the expression of each gene against summary grade. The variable _summarygrade_ is a binarized form of the grade of cancer.
```{R}
lateStagePheno = phenoData(GSE32062.GPL6480_eset)
gradeInfo = data.frame("sample"=row.names(pData(lateStagePheno)),
                       "summarygrade"=pData(lateStagePheno)$summarygrade) 
regressionDataset = lateStageSmall %>% 
  data.frame() %>%
  mutate("sample"=row.names(lateStageSmall)) %>%
  inner_join(gradeInfo,by="sample")

```
To get node weights, we regress gene expression against summary grade and take the resulting p-values. Note that we take the raw p-values and not the FDR, as the conversion to node scores in wActNet provides FDR control.

```{R nodeRegressions}
# BiocManager::install("limma")

library(limma)
design_matrix = model.matrix( ~ factor(summarygrade),data=regressionDataset)
limma_res = regressionDataset %>% select(-c(sample,summarygrade)) %>% t() %>%
  lmFit(object = ., design = design_matrix) %>%
  eBayes()
node_table = limma_res %>%
  topTable(number = Inf) %>%
  data.frame()  
node_p_vals = node_table$P.Value
names(node_p_vals) = rownames(node_table)
```

### Running wActNet

#### Converting p-values to node scores

The `uniform.beta.node.edge.score` function is provided in `wActNet` for converting the list of p-values to scores, in which a higher score corresponds to a lower p-value and vice-versa. FDR control is implemented such that negative scores correspond to p-values above the FDR threshold and positive scores correspond to p-values that are FDR-significant (i.e., at or below the FDR threshold). 

Please note: “Warning: One or both parameters are on the limit of the defined parameter space” indicates that either too many or too few nodes have statistically significant p-values (FDR-adjusted p > 0.05). This suggests that rather than the null uniform-beta mixture distribution, the actual distribution is too close to either the uniform or the beta distribution alone. The user should expand the grid of possible values to be closer to the boundary, using the optim.method, optim.lower, optim.upper, optim.control, and optim.hessian parameters of the uniform.beta.node.edge.score() function. 

```{R}
FDR.node=0.1
FDR.edge=0.05
networkPrep = uniform.beta.node.edge.score(pval.node = node_p_vals, pval.edge = edge_p_vals, FDR.node = FDR.node, FDR.edge = FDR.edge, dat = lateStageSmall, optim.method="Nelder-Mead", optim.lower = -Inf, optim.upper = Inf, optim.control = list(), optim.hessian = FALSE)

hist(node_p_vals, breaks = 50)

#check names of pvals
#names(node_p_vals)
#names(edge_p_vals)

```

#### Running module detection

The `MultiModuleFind` function is used to find active subnetworks. It is an iterative algorithm that finds the largest scoring subnetwork, then removes it from the network and repeats the process. The number of modules `ncluster` must be prespecified. As above, please note that "Warning: No positive nodes" indicates that the function completed algorithm execution before reaching `ncluster` and failed to identify any further optimized or active subnetworks. Here, as with the simulated data above, we compare our proposed method based on both node scores and edge scores with Dittrich's method using node scores only.

```{R}
#all possible optimized subnetworks in simulated network using Dittrich's method using node scores only
modules.n = MultiModuleFind(network = networkPrep$Network,
                          node.scores = networkPrep$NodeScore,
                          edge.scores = networkPrep$EdgeScore,
                          ncluster = 2,
                          method = "NodeOnly")

#all possible optimized subnetworks in simulated network using our proposed method based on node scores and edge scores
modules.ne = MultiModuleFind(network = networkPrep$Network,
                          node.scores = networkPrep$NodeScore,
                          edge.scores = networkPrep$EdgeScore,
                          ncluster = 2,
                          method = "NodeEdge")



```

### Examine results

```{R}
modules.ne
```
Note that if a null result where  the number of nodes and edges are both zero is returned, you may need to either increase the sample size of genes being used or relax the FDR in order to find active subnetworks. Increasing the sample size of genes will give more possible combinations of subnetwork candidates, and relaxing the FDR will allow subnetworks that are less significant to be detected.

This is a list of edges that shows the connected vertices in the active subnetwork that was detected. Note that the number of nodes is 46, and the number of edges is $46 \choose 2$ = 1035, the maximum number for a network with 46 nodes.

Note that if more than one active subnetwork is detected, that second network's information can be accessed by replacing the "[[1]]" below with "[[2]]", and so on for successive subnetworks.
```{R}
#list of active subnetworks detected using Dittrich's method (node scores only)
modules.n

#list of active subnetworks detected using our proposed method (node scores and edge scores)
modules.ne
```

Explore network information:
```{r}
#node scores
V(modules.ne[[1]])$score

#edge scores
E(modules.ne[[1]])$score

#node degree
igraph::degree(modules.ne[[1]])
```

### Visualize Network Results as above
```{r}
library(visNetwork)
multi.mod.n <- modules.n
dat_node<-multi.mod.n[[1]]
 E(dat_node)$dashes<-ifelse(E(dat_node)$score>0,FALSE,FALSE)
  E(dat_node)$width<-(E(dat_node)$score+1.5)*0.3
  V(dat_node)$color<-ifelse(V(dat_node)$score>0,"#E7298A","grey")
  V(dat_node)$size<-(V(dat_node)$score+11)*0.5
  
  E(dat_node)$color<-ifelse(E(dat_node)$score>0,"#66A61E","grey")
 
  data_node<- toVisNetworkData(dat_node)
  
  visNetwork(nodes = data_node$nodes, edges = data_node$edges, height = "800px")%>% visNodes(font=list(size=20))
```

```{r}
multi.mod.ne <- modules.ne
dat_nodeedge<-multi.mod.ne[[1]]
 E(dat_nodeedge)$dashes<-ifelse(E(dat_nodeedge)$score>0,FALSE,FALSE)
  E(dat_nodeedge)$width<-(E(dat_nodeedge)$score+1.5)*0.3
  V(dat_nodeedge)$color<-ifelse(V(dat_nodeedge)$score>0,"#E7298A","grey")
  V(dat_nodeedge)$size<-(V(dat_nodeedge)$score+11)*0.5
  
  E(dat_nodeedge)$color<-ifelse(E(dat_nodeedge)$score>0,"#66A61E","grey")

  
  data_nodeedge <- toVisNetworkData(dat_nodeedge)
  
  visNetwork(nodes = data_nodeedge$nodes, edges = data_nodeedge$edges, height = "800px")%>% visNodes(font=list(size=20))
```




### Runtime dependence on FDR and Number of Genes

```{r}
end.time <- Sys.time()
time.taken <- end.time - start.time
print(c("FDR node: ", FDR.node))
print(c("FDR edge: ", FDR.edge))
print(c("Number of Genes: ", setSize))
print(c("Runtime: ", time.taken))
print(c("Number of Nodes in Active Subnetwork: ", summary(modules.ne)[,1]))

```

#### Selected runtime examples from testing
The two plots below show the impact of the number of genes and the selected FDR values on the runtime. The first plot highlights that the runtime of specific wActNet functions increases as the user increases the number of input genes.

```{r}
library(tidyverse)
library(ggplot2)

NumGenes <- c(100, 300, 500, 1000)
pval.perm.corr <- c(1, 4, 12, 42)
MultiModuleFind <- c(.1, 2, 5, 196) #minutes

runtime_dat <- as.data.frame(cbind(NumGenes, pval.perm.corr, MultiModuleFind))

runtime_dat %>% pivot_longer(2:3, names_to = "Function") %>% ggplot() + geom_point(mapping = aes(x=NumGenes, y = value, color = Function, size = 3), show.legend = c(color=T,size=F)) + theme(plot.caption = element_text(hjust = 0)) + labs(title = "Runtime vs. Number of Genes for Two wActNet Functions", x = "Number of Genes", y = "Runtime in Minutes", caption = str_wrap("This plot highlights how the runtime of specific wActNet functions increases as the user increases the number of input genes. The `MultiModuleFind` function is an iterative algorithm used to find active subnetworks. The `pval.perm.corr` function performs a permutation test on a correlation network to generate edge p-values.", 100))
```

```{r}
FDR <- c(0.05, 0.1, 0.5)
runtimes100genes <- c(36/60, 36/60, 36/60) #seconds/(60s/min) = min
runtimes500genes <- c(15, 21, 48) #minutes

runtime_dat <- as.data.frame(cbind(FDR, runtimes100genes, runtimes500genes))

runtime_dat %>% pivot_longer(2:3, names_to = "Number of Genes") %>% ggplot() + geom_point(mapping = aes(x=FDR, y = value, color = `Number of Genes`, size = 3), show.legend = c(color=T,size=F)) + theme(plot.caption = element_text(hjust = 0)) + labs(title = "Runtime vs. FDR for full script", x = "FDR", y = "Runtime in Minutes", caption = str_wrap("The second plot shows runtimes of a full script, including Dittrich's method and our proposed method, as well as visualizations for both of their respective network outputs. This plot highlights that relaxing the FDR (which allows more nodes to pass the significance threshold, and therefore creates more complex networks) results in longer runtimes.", 100))
```

## Contact Us / Contribute
This package is new, and any and all suggestions are welcomed. You can use GitHub to raise issues, contribute, or communicate with us about the package:

https://github.com/yuy113/wActNet/

## References

Dittrich, M., Klau, G., Rosenwald, A., Dandekar, T., Müller, T.: Identifying functional modules in protein-protein interaction
networks: an integrated exact approach. Bioinformatics 24(13), 223–231 (2008)

Beisser, D., Klau, G., Dandekar, T., Müller, T., Dittrich, M.: Bionet: an r-package for the functional analysis of biological
networks. Bioinformatics 26(8), 1129–1130 (2010)

Ganzfried BF, Riester M, Haibe-Kains B, Risch T, Tyekucheva S, Jazic I, Wang XV, Ahmadifar M, Birrer M, Parmigiani G, Huttenhower C, Waldron L: curatedOvarianData: Clinically Annotated Data for the Ovarian Cancer Transcriptome. Database, Volume 2013. doi:10.1093/database/bat013, https://academic.oup.com/database/article/doi/10.1093/database/bat013/330978 
